### PR TITLE
[ENHANCEMENT] Improve UI for in-code data contexts

### PR DESCRIPTION
**Pull request by anthonyburdi**
When initializing a DataContext in-memory we have to specify many parameters (e.g. a notebook where we don't have access to a filesystem like [EMR Spark](https://docs.greatexpectations.io/en/latest/guides/how_to_guides/configuring_data_contexts/how_to_instantiate_a_data_context_on_an_emr_spark_cluster.html), [Databricks](https://docs.greatexpectations.io/en/latest/guides/how_to_guides/configuring_data_contexts/how_to_instantiate_a_data_context_on_a_databricks_spark_cluster.html)). For many configurations, most of these parameters are not modified from defaults. This PR proposes a set of defaults to simplify initialization of DataContexts and Datasources via their configuration objects. See also [this design doc](https://github.com/superconductive/design/blob/main/docs/20201020_in_memory_ge.md).

NOTE: This PR also includes code from #2075 - originally I had planned to merge this PR into the feature branch from #2075 as well, but to keep things moving along I'm keeping them separate and this PR is into `develop`. We can merge this into `anthony/enhancement/improve-ui-and-tracking-for-in-memory-config` instead of `develop` or just merge this PR and #2075 directly into develop.

Changes proposed in this pull request:
- Add defaults for DataContextConfig and DatasourceConfig
- Create "BackendEcosystem" classes for typical StoreBackend configurations, e.g. S3BackendEcosystem, FilesystemBackendEcosystem, GCSBackendEcosystem, DatabaseBackendEcosystem
- Define base defaults, and use these in templates.py (this can be improved)
- Test that configuration works as expected using these defaults, including environment variable substitution in configs.

Open questions:
- What should the default `plugins_directory` be? (`None` or `plugins/`)

Note: this branch also includes [anthony/enhancement/improve-ui-and-tracking-for-in-memory-config](https://github.com/great-expectations/great_expectations/tree/anthony/enhancement/improve-ui-and-tracking-for-in-memory-config) which is planned to be merged into develop first.

### CODE REVIEW

File: great_expectations/data_context/types/base.py [hunk 1]
```diff
@@ -437,6 +488,482 @@ def validate_schema(self, data, **kwargs):
...
+        validations_store_name: str = "validations_S3_store",
+        evaluation_parameter_store_name: str = "evaluation_parameter_store",
+    ):
+        # Initialize base defaults
+        super().__init__()
+
+        # Use default_bucket_name if separate store buckets are not provided
+        expectations_store_bucket_name = (
+            expectations_store_bucket_name
+            if expectations_store_bucket_name is not None
```
Comment (talagluck) [inline great_expectations/data_context/types/base.py]: `
```suggestion
            if expectations_store_bucket_name
```

File: great_expectations/data_context/types/base.py [hunk 2]
```diff
@@ -437,6 +488,482 @@ def validate_schema(self, data, **kwargs):
...
+            else default_bucket_name
+        )
+        validations_store_bucket_name = (
+            validations_store_bucket_name
+            if validations_store_bucket_name is not None
```
Comment (talagluck) [inline great_expectations/data_context/types/base.py]:
```suggestion
            if validations_store_bucket_name
```

File: great_expectations/data_context/types/base.py [hunk 3]
```diff
@@ -437,6 +488,482 @@ def validate_schema(self, data, **kwargs):
...
+            else default_bucket_name
+        )
+        data_docs_bucket_name = (
+            data_docs_bucket_name
+            if data_docs_bucket_name is not None
```
Comment (talagluck) [inline great_expectations/data_context/types/base.py]:
```suggestion
            if data_docs_bucket_name
```

File: great_expectations/data_context/types/base.py [hunk 4]
```diff
@@ -437,6 +488,482 @@ def validate_schema(self, data, **kwargs):
...
+        validations_store_name: str = "validations_GCS_store",
+        evaluation_parameter_store_name: str = "evaluation_parameter_store",
+    ):
+        # Initialize base defaults
+        super().__init__()
+
+        # Use default_bucket_name if separate store buckets are not provided
+        expectations_store_bucket_name = (
+            expectations_store_bucket_name
+            if expectations_store_bucket_name is not None
```
Comment (talagluck) [inline great_expectations/data_context/types/base.py]:
```suggestion
            if expectations_store_bucket_name
```

File: great_expectations/data_context/types/base.py [hunk 5]
```diff
@@ -437,6 +488,482 @@ def validate_schema(self, data, **kwargs):
...
+            else default_bucket_name
+        )
+        validations_store_bucket_name = (
+            validations_store_bucket_name
+            if validations_store_bucket_name is not None
```
Comment (talagluck) [inline great_expectations/data_context/types/base.py]:
```suggestion
            if validations_store_bucket_name
```

File: great_expectations/data_context/types/base.py [hunk 6]
```diff
@@ -437,6 +488,482 @@ def validate_schema(self, data, **kwargs):
...
+            else default_bucket_name
+        )
+        data_docs_bucket_name = (
+            data_docs_bucket_name
+            if data_docs_bucket_name is not None
```
Comment (talagluck) [inline great_expectations/data_context/types/base.py]:
```suggestion
            if data_docs_bucket_name
```

File: great_expectations/data_context/types/base.py [hunk 7]
```diff
@@ -437,6 +488,482 @@ def validate_schema(self, data, **kwargs):
...
+            else default_bucket_name
+        )
+
+        # Use default_project_name if separate store projects are not provided
+        expectations_store_project_name = (
+            expectations_store_project_name
+            if expectations_store_project_name is not None
```
Comment (talagluck) [inline great_expectations/data_context/types/base.py]:
```suggestion
            if expectations_store_project_name
```

File: great_expectations/data_context/types/base.py [hunk 8]
```diff
@@ -437,6 +488,482 @@ def validate_schema(self, data, **kwargs):
...
+            else default_project_name
+        )
+        validations_store_project_name = (
+            validations_store_project_name
+            if validations_store_project_name is not None
```
Comment (talagluck) [inline great_expectations/data_context/types/base.py]:
```suggestion
            if validations_store_project_name
```

File: great_expectations/data_context/types/base.py [hunk 9]
```diff
@@ -437,6 +488,482 @@ def validate_schema(self, data, **kwargs):
...
+            else default_project_name
+        )
+        data_docs_project_name = (
+            data_docs_project_name
+            if data_docs_project_name is not None
```
Comment (talagluck) [inline great_expectations/data_context/types/base.py]:
```suggestion
            if data_docs_project_name
```

File: great_expectations/data_context/types/base.py [hunk 10]
```diff
@@ -437,6 +488,482 @@ def validate_schema(self, data, **kwargs):
...
+        validations_store_name: str = "validations_database_store",
+        evaluation_parameter_store_name: str = "evaluation_parameter_store",
+    ):
+        # Initialize base defaults
+        super().__init__()
+
+        # Use default credentials if seprate credentials not supplied for expectations_store and validations_store
+        expectations_store_credentials = (
+            expectations_store_credentials
+            if expectations_store_credentials is not None
```
Comment (talagluck) [inline great_expectations/data_context/types/base.py]:
```suggestion
            if expectations_store_credentials
```

File: great_expectations/data_context/types/base.py [hunk 11]
```diff
@@ -437,6 +488,482 @@ def validate_schema(self, data, **kwargs):
...
+            else default_credentials
+        )
+        validations_store_credentials = (
+            validations_store_credentials
+            if validations_store_credentials is not None
```
Comment (talagluck) [inline great_expectations/data_context/types/base.py]:
```suggestion
            if validations_store_credentials
```

File: great_expectations/data_context/types/base.py [hunk 12]
```diff
@@ -437,6 +488,482 @@ def validate_schema(self, data, **kwargs):
...
+        #  for this directory when it is used and graceful handling if not found?
+
+        # Set defaults via backend_ecosystem if one is passed in
+        # What happens / should happen if a user specifies a backend_ecosystem
+        #  and also items that would be specified inside of it e.g. expectations_store_name?
+        # Override attributes from backend_ecosystem with any items passed into the constructor:
+        if backend_ecosystem:
+            config_version = (
+                config_version
+                if config_version is not None
```
Comment (talagluck) [inline great_expectations/data_context/types/base.py]:
```suggestion
                if config_version
```

File: great_expectations/data_context/types/base.py [hunk 13]
```diff
@@ -437,6 +488,482 @@ def validate_schema(self, data, **kwargs):
...
+                else backend_ecosystem.config_version
+            )
+            stores = stores if stores is not None else backend_ecosystem.stores
+            expectations_store_name = (
+                expectations_store_name
+                if expectations_store_name is not None
```
Comment (talagluck) [inline great_expectations/data_context/types/base.py]:
```suggestion
                if expectations_store_name
```

File: great_expectations/data_context/types/base.py [hunk 14]
```diff
@@ -437,6 +488,482 @@ def validate_schema(self, data, **kwargs):
...
+                else backend_ecosystem.expectations_store_name
+            )
+            validations_store_name = (
+                validations_store_name
+                if validations_store_name is not None
```
Comment (talagluck) [inline great_expectations/data_context/types/base.py]:
```suggestion
                if validations_store_name
```

File: great_expectations/data_context/types/base.py [hunk 15]
```diff
@@ -437,6 +488,482 @@ def validate_schema(self, data, **kwargs):
...
+                else backend_ecosystem.validations_store_name
+            )
+            evaluation_parameter_store_name = (
+                evaluation_parameter_store_name
+                if evaluation_parameter_store_name is not None
```
Comment (talagluck) [inline great_expectations/data_context/types/base.py]:
```suggestion
                if evaluation_parameter_store_name
```

File: great_expectations/data_context/types/base.py [hunk 16]
```diff
@@ -437,6 +488,482 @@ def validate_schema(self, data, **kwargs):
...
+                else backend_ecosystem.evaluation_parameter_store_name
+            )
+            validation_operators = (
+                validation_operators
+                if validation_operators is not None
```
Comment (talagluck) [inline great_expectations/data_context/types/base.py]:
```suggestion
                if validation_operators
```

File: great_expectations/data_context/types/base.py [hunk 17]
```diff
@@ -437,6 +488,482 @@ def validate_schema(self, data, **kwargs):
...
+                else backend_ecosystem.validation_operators
+            )
+            data_docs_sites = (
+                data_docs_sites
+                if data_docs_sites is not None
```
Comment (talagluck) [inline great_expectations/data_context/types/base.py]:
```suggestion
                if data_docs_sites
```

File: great_expectations/data_context/types/base.py [hunk 18]
```diff
@@ -437,6 +488,482 @@ def validate_schema(self, data, **kwargs):
...
+                else backend_ecosystem.data_docs_sites
+            )
+
+        if commented_map is None:
```
Comment (talagluck) [inline great_expectations/data_context/types/base.py:1014]:
```suggestion
        if not commented_map:
```

File: great_expectations/data_context/types/base.py [hunk 19]
```diff
@@ -437,6 +488,482 @@ def validate_schema(self, data, **kwargs):
...
+            commented_map = CommentedMap()
+        self._commented_map = commented_map
+        self._config_version = config_version
+        if datasources is None:
```
Comment (talagluck) [inline great_expectations/data_context/types/base.py:1018]:
```suggestion
        if not datasources:
```

File: great_expectations/data_context/types/base.py [hunk 20]
```diff
@@ -437,6 +488,482 @@ def validate_schema(self, data, **kwargs):
...
+        if not isinstance(validation_operators, dict):
+            raise ValueError(
+                "validation_operators must be configured with a dictionary"
+            )
+        self.validation_operators = validation_operators
+        self.stores = stores
+        self.notebooks = notebooks
+        self.data_docs_sites = data_docs_sites
+        self.config_variables_file_path = config_variables_file_path
+        if anonymous_usage_statistics is None:
```
Comment (talagluck) [inline great_expectations/data_context/types/base.py:1034]:
```suggestion
        if not anonymous_usage_statistics:
```

File: great_expectations/data_context/templates.py [hunk 21]
```diff
@@ -1,7 +1,36 @@
...
+
+
+class YAMLToString(YAML):
+    """
+    Get yaml dump as a string: https://yaml.readthedocs.io/en/latest/example.html#output-of-dump-as-a-string
+    """
+
+    def dump(self, data, stream=None, **kw):
+        inefficient = False
+        if stream is None:
```
Comment (talagluck) [inline great_expectations/data_context/templates.py]:
```suggestion
        if not stream:
```

File: great_expectations/data_context/types/base.py [hunk 22]
```diff
@@ -29,77 +32,120 @@
...
             )
-            raise
 
-    def to_yaml(self, outfile):
-        commented_map = deepcopy(self.commented_map)
-        commented_map.update(dataContextConfigSchema.dump(self))
-        yaml.dump(commented_map, outfile)
+        data_asset_type = (
+            defaults_class.data_asset_type()
+            if data_asset_type is None
```
Comment (talagluck) [inline great_expectations/data_context/types/base.py]:
```suggestion
            if not data_asset_type
```

File: great_expectations/data_context/types/base.py [hunk 23]
```diff
@@ -29,77 +32,120 @@
...
+            else data_asset_type
+        )
+        module_name = (
+            defaults_class.module_name() if module_name is None else module_name
```
Comment (talagluck) [inline great_expectations/data_context/types/base.py]:
```suggestion
            defaults_class.module_name() if not module_name else module_name
```

**Review by talagluck — COMMENTED** [review]
Overall looks great! Just some totally optional stylistic suggestions.

**Comment by anthonyburdi** [comment]
Note: @jcampbell approved the design of these changes in a review on Mon Nov 30 2020

**Review by talagluck — APPROVED** [review]
LGTM!

**Pull request merged**